### PR TITLE
✨ Add parameter IO support for more formats supported by pandas

### PR DESCRIPTION
**Pull request by patrickhaetti**
### Change summary
- added xlsx & ods parameter:
     - added xlsx.py in builtin/io/pandas. Replace inf / -inf with NaN when saving and replace NaN with inf in 'maximum' and -inf in      'minimum 'column of Dataframe  #888
     - setup.cfg: add ```xlsx = glotaran.builtin.io.pandas.xlsx```
     - added test_pandas_parameters.py and add 'data' subfolder which includes parameter.xlsx and parameter.yaml 
     - added tests with parameter files *.xlsx and *.ods
     - added ```openpyxl>=3.0.9``` to setup.cfg / install_requires
     - added ```odfpy>=1.4.1``` to setup.cfg / install_requires
     - add ```openpyxl==3.0.9``` to requirements_dev.txt to read/write xlsx

- Added test files to pandas/tests/data with parameter.yaml and create  csv, tsv, xlsx parameter files including one type of each parameter definition.

- added tsv parameter:
     - add tsv.py in builtin/io/pandas for tab-separated-value parameter
     - setup.cfg: add ```tsv = glotaran.builtin.io.pandas.tsv```
     - added tsv.py to builtin/io/pandas.
     - added parameter.tsv in test/data

### Checklist

- [x] ✔️ Passing the tests (mandatory for all PR's)
- [x] 👌 Closes issue (mandatory for ✨ feature and 🩹 bug fix PR's)
- [x] 🧪 Adds new tests for the feature (mandatory for ✨ feature and 🩹 bug fix PR's)
- [ ] 📚 Adds documentation of the feature

### CODE REVIEW

File: glotaran/builtin/io/pandas/test/data/parameter.yaml
Comment (s-weigand) [inline glotaran/builtin/io/pandas/test/data/parameter.yaml:29]: For testing one type of each parameter definition should be enough:
- Short list style no options
```yaml
pure_list: [1.0, 2.0]
```
- Short list style with options
```yaml
list_with_options: [3.0, 4.0, {vary: False}]
```
- Verbose list style with and without options
```yaml
verbose_list: 
  - ["all_defaults", 5.0]
  - ["no_defaults", 6.0, {non-negative: True, vary: False, minimum: -1, maximum: 1}]
  - ["expression_only", {expr: $verbose_list.all_defaults + $verbose_list.no_defaults}]
```

This will also need to recreate the excel file.

File: glotaran/builtin/io/pandas/test/test_pandas_parameters.py
Comment (s-weigand) [inline glotaran/builtin/io/pandas/test/test_pandas_parameters.py:10]: This is highly dependent on where you execute pytest, better use the location of the current file.
```suggestion
PANDAS_TEST_DATA = Path(__file__).parent / "data"
PATH_XLSX = PANDAS_TEST_DATA  / "parameter.xlsx"
PATH_YAML = PANDAS_TEST_DATA  / "parameter.yaml"
```

Since `DATA_PATH` is only used in `test_save_parameters_xlsx`, it can be omitted.

File: glotaran/builtin/io/pandas/test/test_pandas_parameters.py
Comment (s-weigand) [inline glotaran/builtin/io/pandas/test/test_pandas_parameters.py:26]:
```suggestion
    parameter_path = tmp_path / "testparameters.xlsx"
```
See other suggestion in this file.

File: requirements_dev.txt
Comment (s-weigand) [inline requirements_dev.txt:10]:
```suggestion
numpy==1.20.3
openpyxl==3.0.9
```
I guess removing `numpy` wasn't intended. 😉

**Review by s-weigand — CHANGES_REQUESTED** [review]
Thanks for the contribution, but there are some changes needed (see review).

In addition: 
- It would also be useful to have a `*.csv` file for testing that it looks like it should.
- Having an option for writing to `csv` e.g. `strip_default=True` would be useful when `csv` files are edited in excel.
- Please also add a plugin for `tsv` (tab separated values) the only difference to `csv` should be  `sep=+\t` in the `pandas` method.



@jsnel  @joernweissenborn 
- What do you think about adding `openpyxl` as a runtime dependency?
- What do you think about setting all default parameter values to empty strings, this would reduce cognitive load and make parameter files easier to read.

**Comment by joernweissenborn** [comment]
> Thanks for the contribution, but there are some changes needed (see review).
> 
> In addition:
> 
> * It would also be useful to have a `*.csv` file for testing that it looks like it should.
> * Having an option for writing to `csv` e.g. `strip_default=True` would be useful when `csv` files are edited in excel (to reduce code duplication best factor the replacing out in a separate function which then can be unit tested).
> * Please also add a plugin for `tsv` (tab separated values) the only difference to `csv` should be  `sep=+\t` in the `pandas` method.
> 
> @jsnel @joernweissenborn
> 
> * What do you think about adding `openpyxl` as a runtime dependency?
> * What do you think about setting all default parameter values to empty strings, this would reduce cognitive load and make parameter files easier to read.

I am ok with both.

**Comment by jsnel** [comment]
> * What do you think about adding `openpyxl` as a runtime dependency?
> * What do you think about setting all default parameter values to empty strings, this would reduce cognitive load and make parameter files easier to read.

Yes to both.

**Comment by joernweissenborn** [comment]
@patrickhaetti Looking at the commit mess I suggest you close this PR and and open a new one, with clean commit history. I think @glotaran/maintainers would also prefer it.

**Comment by s-weigand** [comment]
> 
> 
> @patrickhaetti Looking at the commit mess I suggest you close this PR and and open a new one, with clean commit history. I think @glotaran/maintainers would also prefer it.

Let's keep the PR but I will remove the mess 😉

**Comment by s-weigand** [comment]
@patrickhaetti I cleaned up the commit history please make sure to reset your local branch to the remote state.

E.g. if your remote is called `origin`
```console
git reset --hard orgin/create-xlsx-parameter
```

If you have uncommitted changes just stash them before (`git stash`) and pop the stash afterward (`git stash pop`).
If you have local commits you didn't push yet just create a branch at the current state(`git branch save-xlsx-branch`) and cherry-pick the commits over after you reset to the remote branch.

**Comment by patrickhaetti** [comment]
@s-weigand  thank you for cleaning up, looks way more better now

**Comment by s-weigand** [comment]
I will implement the `strip_default=True` functionality on the parameter level, since we will also need this for `yaml` round trips.

**Comment by jsnel** [comment]
In this PR result comparison was failing on a singular point in the residual. THis is not a problem with this PR but reason to further improve the github actions that does the results comparison, see issue  #937 which will be fixed in addressed in #937. 

<details>
<summary>FAILED .github/test_result_consistency.py::test_result_data_var_consistency[study_transient_absorption_target_analysis-residual]</summary>

E       AssertionError: Result data_var data mismatch: 'residual' in 'dataset1.nc'.
E         With sum of absolute difference: 2.1183739743033033e-07 and shape: (335, 346)
E         Mean difference: 1.827602428007336e-12
E         
E       assert False
E        +  where False = <function allclose.<locals>._allclose at 0x7efc69aaa280>(<xarray.DataArray 'residual' (time: 335, spectral: 346)>\narray([[-4.030000e-03,  1.100000e-04, -6.014408e-15, ...,  2....95 0.02343 0.04457 ... 869.9 889.9 909.9\n  * spectral  (spectral) float64 378.0 379.4 380.9 382.3 ... 851.8 853.1 854.4, <xarray.DataArray 'residual' (time: 335, spectral: 346)>\narray([[-4.030000e-03,  1.100000e-04, -6.017322e-15, ...,  2....95 0.02343 0.04457 ... 869.9 889.9 909.9\n  * spectral  (spectral) float64 378.0 379.4 380.9 382.3 ... 851.8 853.1 854.4, atol=array([[1.934e-10, 1.934e-10, 1.934e-10, ..., 1.934e-10, 1.934e-10,\n        1.934e-10],\n       [1.934e-10, 1.934e-10, ...1.934e-10,\n        1.934e-10],\n       [1.934e-10, 1.934e-10, 1.934e-10, ..., 1.934e-10, 1.934e-10,\n        1.934e-10]]), rtol=1e-05, print_fail=20)

.github/test_result_consistency.py:221: AssertionError
----------------------------- Captured stdout call -----------------------------
allclose first 1 failures:
  (87, 344): 4.579833638334098e-06 4.579591726559538e-06
=========================== short test summary info ============================
FAILED .github/test_result_consistency.py::test_result_data_var_consistency[study_transient_absorption_target_analysis-residual]
======================== 1 failed, 296 passed in 5.59s =========================
Error: Process completed with exit code 1.

</details>

**Comment by jsnel** [comment]
Thank you for this PR @patrickhaetti. Nice work! 

Have all your comments been addressed @s-weigand? If so I'd like to schedule it in the merge pipeline after #923 is merged in.

**Comment by patrickhaetti** [comment]
> Thank you for this PR @patrickhaetti. Nice work!
> 
> Have all your comments been addressed @s-weigand? If so I'd like to schedule it in the merge pipeline after #923 is merged in.

it was a pleasure!

File: glotaran/builtin/io/pandas/tsv.py
Comment (s-weigand) [inline glotaran/builtin/io/pandas/tsv.py:14]:
```suggestion
    """Plugin for TSV data io."""
```

File: glotaran/builtin/io/pandas/xlsx.py
Comment (s-weigand) [inline glotaran/builtin/io/pandas/xlsx.py:31]: This should also use `glotaran.utils.io.safe_parameters_fillna`.

File: glotaran/builtin/io/pandas/xlsx.py
Comment (s-weigand) [inline glotaran/builtin/io/pandas/xlsx.py:46]: This should also use `glotaran.utils.io.safe_parameters_replace`

File: glotaran/utils/io.py
Comment (s-weigand) [inline glotaran/utils/io.py:207]: This function needs tests.

File: glotaran/utils/io.py
Comment (s-weigand) [inline glotaran/utils/io.py:191]: This function needs tests.

**Review by s-weigand — CHANGES_REQUESTED** [review]
@patrickhaetti Please fix those minor issues.

**Review by jsnel — DISMISSED** [review]
It looks good to me. But @s-weigand will have to rebase prior to merging and check compatibility with recent changes to API related to project/result saving (and perhaps add the extra tests your requested yourself ^^)

File: glotaran/utils/io.py
Comment (jsnel) [reply glotaran/utils/io.py:229]: Added by @patrickhaetti in 7233dbce5540b769a520b8e648b3be669e5deafb

File: glotaran/utils/io.py
Comment (jsnel) [reply glotaran/utils/io.py:213]: Added by @patrickhaetti in 7233dbce5540b769a520b8e648b3be669e5deafb

File: glotaran/builtin/io/pandas/xlsx.py
Comment (jsnel) [inline glotaran/builtin/io/pandas/xlsx.py:16]: Technically, specifically the xlsx (the Office Open XML standard) and ods (the Open Document Standard) format are not specific to Excel, but open standards. So the class could be named differently, but since it uses the read_excel function from pandas under the hood, who are we to change the name?  
Note that the pandas `read_excel` function supports xls, xlsx, xlsm, xlsb, odf, ods and odt, so more than just xlsx and ods.

File: glotaran/builtin/io/pandas/test/test_pandas_parameters.py
Comment (jsnel) [inline glotaran/builtin/io/pandas/test/test_pandas_parameters.py:30]: Somewhat counter intuitive, you could change the format_name to `format_name="ods"` and it would still  work (unit tests still pass), because it would select the plugin based on format name, but because that's the same plugin as used for 'xlsx' saving the file still works. Under the hood pandas.to_excel is used which by itself infers how to save it based on the file extension, the format_name information is not propagated. 

Not something I expect to be an issue, but important to note.

**Review by jsnel — APPROVED** [review]
LGTM ;)

**Review by jsnel — DISMISSED** [review]
Looks like it's finally ready to be merged in, sorry for the delay @patrickhaetti and thanks for all the work you put into it!

**Review by s-weigand — APPROVED** [review]
@patrickhaetti I refactored the tests to have 100% coverage and less code duplication.
I also added the optional keyword-only argument `as_optimized` to the `save_parameters` methods which can be passed to `ParameterGroup.to_dataframe` to omit values that are only resent after an optimization.

All looks good to me now, so we are ready to merge 🚀

File: glotaran/deprecation/modules/test/test_parameter_parameter_group.py
Comment (joernweissenborn) [inline glotaran/deprecation/modules/test/test_parameter_parameter_group.py:18]: @s-weigand Does this correctly translate to `(-)np.inf`? it is important to have them lie this in optimization if no value is provided.

**Review by joernweissenborn — COMMENTED** [review]
@s-weigand Only one Q for before my lgtm, it is in the review.

File: glotaran/deprecation/modules/test/test_parameter_parameter_group.py
Comment (s-weigand) [reply glotaran/deprecation/modules/test/test_parameter_parameter_group.py:18]: Yap it is tested in `glotaran/utils/test/test_io.py`. 
But maybe we should make dropping the `inf` values optional for text-based formats (`csv`, `tsv`)?

Comment (s-weigand) [reply glotaran/deprecation/modules/test/test_parameter_parameter_group.py:18]: Also `tsv` and `csv` aren't consistent ... in that 😖

**Review by joernweissenborn — APPROVED** [review]
lgtm

File: glotaran/deprecation/modules/test/test_parameter_parameter_group.py
Comment (s-weigand) [reply glotaran/deprecation/modules/test/test_parameter_parameter_group.py:18]: Fixed consistency issue and added test for mixed loading in 37bb74a

Comment (s-weigand) [reply glotaran/deprecation/modules/test/test_parameter_parameter_group.py]: Restored original test 9602f53, so now the deprecate method behaves exactly like before

**Review by s-weigand — APPROVED** [review]

**Comment by patrickhaetti** [comment]
Hi @all, I am happy to hear the merge happened!

**Pull request merged**
